### PR TITLE
capi: Add separate dependency scripts

### LINKED ISSUE #78: Support running image-builder in RHEL 7
The Makefile in images/capi/packer/ami and images/capi/packer/gce uses `shasum` to verify the downloaded packer plugins.

In some distributions, such as RHEL 7, `shasum` is actually `sha256sum`. I think we can update the Makefile to detect whether `shasum` or `sha256sum` is available on the machine. If neither are there, exit with an error.

### PR DESCRIPTION
**Pull request by codenrhoden**
This patch splits apart the existing hack/image-tools.sh script into
separate ones that can be called via Make targets. This allows a user to
use `make` to ensure their system has dependencies installed for building
images for their specific provider.

The Makefile is enhanced into sections for Building, Cleaning, and
Dependencies.

All the scripts have passed `shellcheck`, and have been tested on MacOS and Linux (CentOS 7). The previous `image-tools.sh` was clearly never tested on Linux, because it used a non-existing argument (`--ignore-missing`) to `sha256sum`, causing the script to fail when used. So instead we use `sed` to remove the lines we don't need when validating the packer checksum.

fixes #78 
supersedes #140 

This patch also groups the makefile targets into categories. The new `help` output looks like:
```sh
Usage:
  make <target>

Helpers
  help                         Display this help

Dependencies
  deps                         Installs/checks all dependencies
  deps-ami                     Installs/checks dependencies for AMI builds
  deps-azure                   Installs/checks dependencies for Azure builds
  deps-do                      Installs/checks dependencies for DigitalOcean builds
  deps-gce                     Installs/checks dependencies for GCE builds
  deps-ova                     Installs/checks dependencies for OVA builds

Builds
  build-ami-default            Builds the AMI default image
  build-azure-sig-ubuntu-1804  Builds Ubuntu 18.04 Azure managed image in Shared Image Gallery
  build-azure-vhd-ubuntu-1804  Builds Ubuntu 18.04 VHD image for Azure
  build-do-default             Builds the DigitalOcean snapshot default image
  build-gce-default            Builds the GCE default image
  build-ova-centos-7           Builds CentOS 7 OVA w local hypervisor
  build-ova-photon-3           Builds Photon 3 OVA w local hypervisor
  build-ova-ubuntu-1804        Builds Ubuntu 18.04 OVA w local hypervisor
  build-ova-ubuntu-2004        Builds Ubuntu 20.04 OVA w local hypervisor
  build-esx-centos-7           Builds CentOS 7 OVA w remote hypervisor
  build-esx-photon-3           Builds Photon 3 OVA w remote hypervisor
  build-esx-ubuntu-1804        Builds Ubuntu 18.04 OVA w remote hypervisor
  build-esx-ubuntu-2004        Builds Ubuntu 20.04 OVA w remote hypervisor

Cleaning
  clean                        Cleans all local image caches
  clean-ova                    Removes all ova image output directories (see NOTE at top of help)
  clean-packer-cache           Removes the packer cache
```

/assign @detiber

### CODE REVIEW

File: images/capi/hack/ensure-ansible.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+set -o nounset
+set -o pipefail
+
+_version="2.8.0"
+
+# Change directories to the parent directory of the one in which this
+# script is located.
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+
+if command -v ansible >/dev/null 2>&1; then exit 0; fi
```
Comment (detiber) [inline images/capi/hack/ensure-ansible.sh]: Do we also need to verify that the version of the existing ansible binary is >= the version we expect?

File: images/capi/hack/ensure-goss.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+source hack/utils.sh
+
+_version="0.3.0"
```
Comment (detiber) [inline images/capi/hack/ensure-goss.sh]: Probably not overly important here, but might be good to track that we should look into updating this to v1.0.0: https://github.com/YaleUniversity/packer-provisioner-goss/releases/tag/v1.0.0

File: images/capi/hack/ensure-ansible.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+_version="2.8.0"
```
Comment (detiber) [inline images/capi/hack/ensure-ansible.sh]: Might be good to track that we should look into potentially updating this to the latest 2.8.x release or possibly 2.9.x

File: images/capi/hack/ensure-goss.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+darwin_sha256="7ae43b5dbd26a166c8673fc7299e91d1c2244c7d2b3b558ce04e2e53acfa6f88"
+linux_sha256="28be39d0ddf9ad9c14e432818261abed2f2bd83257cfba213e19d5c59b710d03"
+_bin_url="https://github.com/YaleUniversity/packer-provisioner-goss/releases/download/v${_version}/packer-provisioner-goss-v${_version}-${HOSTOS}-${HOSTARCH}"
+
+_binfile="${HOME}/.packer.d/plugins/packer-provisioner-goss"
+if [ -f "${_binfile}" ]; then
+  { [ -x "${_binfile}" ] && exit 0; } || rm -f "${_binfile}"
```
Comment (detiber) [inline images/capi/hack/ensure-goss.sh]: Can we attempt to verify version of the binary here as well?

File: images/capi/hack/ensure-jq.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+_version="1.6" # earlier versions don't follow the same OS/ARCH patterns
+
+# Change directories to the parent directory of the one in which this
+# script is located.
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+
+source hack/utils.sh
+
+if command -v jq >/dev/null 2>&1; then exit 0; fi
```
Comment (detiber) [inline images/capi/hack/ensure-jq.sh]: Should we verify version here?

File: images/capi/hack/ensure-packer.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+_version="1.4.1"
```
Comment (detiber) [inline images/capi/hack/ensure-packer.sh]: Should we track investigating updating this to the latest 1.4.x, or possibly 1.5.x?

File: images/capi/hack/ensure-packer.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+# Change directories to the parent directory of the one in which this
+# script is located.
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+
+source hack/utils.sh
+
+if command -v packer >/dev/null 2>&1; then exit 0; fi
```
Comment (detiber) [inline images/capi/hack/ensure-packer.sh]: Should we also check version is the expected version here as well?

**Comment by codenrhoden** [comment]
@detiber Agree with all your feedback. Would you be okay with me filing issues to implement those items? My goal here was to get it started and enforce what we had, making as few changes as possible to what was in `image-tools.sh`.

**Comment by detiber** [comment]
@codenrhoden +1 to followup issues
/lgtm

**Pull request merged**